pgrx-sql-entity-graph/src/pg_extern/returning.rs

```diff
@@ -111,62 +102,40 @@ impl Returning {
-                                            )
-                                        })? {
-                                            GenericArgument::Type(Type::Path(this_path)) => {
-                                                segments = this_path.path.segments.clone();
-                                                saw_option_ident = true;
-                                                found_option = true;
-                                                continue 'outer;
-                                            }
-                                            _ => continue,
-                                        };
-                                    }
-                                    _ => continue,
-                                },
-                                "SetOfIterator" => saw_setof_iterator = true,
-                                "TableIterator" => {
-                                    if found_option {
-                                        segments =
-                                            Punctuated::from_iter(std::iter::once(segment.clone()));
-                                        found_option = false;
-                                        continue 'outer;
-                                    }
-                                    saw_table_iterator = true
-                                }
-                                _ => (),
+
+                    loop {
+                        if let Some(segment) = segments.filter_last_ident("Option") {
+                            let PathArguments::AngleBracketed(generics) = &segment.arguments else {
+                                unreachable!()
                             };
+                            let Some(GenericArgument::Type(Type::Path(this_path))) =
+                                generics.args.last()
+                            else {
+                                return Err(syn::Error::new_spanned(
+                                    generics,
+                                    "where's the generic args?",
+                                ));
+                            };
+                            is_option = true;
+                            segments = this_path.path.segments.clone(); // recurse deeper
+                        } else {
+                            if segments.last_ident_is("SetOfIterator") {
+                                is_setof_iter = true;
+                            } else if segments.last_ident_is("TableIterator") {
+                                is_table_iter = true;
+                            }
+                            break;
                         }
-                        break;
                     }
 
-                    if saw_setof_iterator {
+                    if is_setof_iter {
                         let last_path_segment = option_inner_path.segments.last();
-                        let (used_ty, optional) = match &last_path_segment.map(|ps| &ps.arguments) {
```

Comment:
the destructuring of the tuple here always used the same value every time, so...